File: 3. Behavioral Design Patterns/Template Method/Javascript/TemplateMethodExample.js
// Template Method Pattern Example in JavaScript
// DataProcessor with CSV and JSON implementations

class DataProcessor {
  processData() {
    this.loadData()
    this.processDataImpl()
    this.saveData()
  }
  loadData() { throw new Error('Abstract method') }
  processDataImpl() { throw new Error('Abstract method') }
  saveData() { throw new Error('Abstract method') }
}

class CSVProcessor extends DataProcessor {
  loadData() { console.log('Loading CSV data') }
  processDataImpl() { console.log('Processing CSV data') }
  saveData() { console.log('Saving CSV data') }
}

class JSONProcessor extends DataProcessor {
  loadData() { console.log('Loading JSON data') }
  processDataImpl() { console.log('Processing JSON data') }
  saveData() { console.log('Saving JSON data') }
}

;(function main() {
  const csv = new CSVProcessor()
  csv.processData()
  console.log()
  const json = new JSONProcessor()
  json.processData()
})()

File: 3. Behavioral Design Patterns/Template Method/Kotlin/TemplateMethodExample.kt
// Template Method Pattern Example in Kotlin
// DataProcessor with CSV and JSON implementations

abstract class DataProcessor {
    fun processData() {
        loadData()
        processDataImpl()
        saveData()
    }
    protected abstract fun loadData()
    protected abstract fun processDataImpl()
    protected abstract fun saveData()
}

class CSVProcessor : DataProcessor() {
    override fun loadData() = println("Loading CSV data")
    override fun processDataImpl() = println("Processing CSV data")
    override fun saveData() = println("Saving CSV data")
}

class JSONProcessor : DataProcessor() {
    override fun loadData() = println("Loading JSON data")
    override fun processDataImpl() = println("Processing JSON data")
    override fun saveData() = println("Saving JSON data")
}

fun main() {
    val csv: DataProcessor = CSVProcessor()
    csv.processData()
    println()
    val json: DataProcessor = JSONProcessor()
    json.processData()
}
